File: agent/core/AutonomousAgent.js
require('dotenv').config();
const path = require('path');
const Anthropic = require('@anthropic-ai/sdk');

// Import core components
const StateManager = require('../state/StateManager');
const DecisionEngine = require('../decision/DecisionEngine');
const ErrorHandler = require('../error/ErrorHandler');

// Import tools
const SearchTool = require('../tools/SearchTool');
const AnalyzeTool = require('../tools/AnalyzeTool');
const OrganizeTool = require('../tools/OrganizeTool');
const TransformTool = require('../tools/TransformTool');
const TranslateTool = require('../tools/TranslateTool');
const LocaleTool = require('../tools/LocaleTool');
const SetupTool = require('../tools/SetupTool');
const IntegrateTool = require('../tools/IntegrateTool');
const ValidateTool = require('../tools/ValidateTool');
const TestTool = require('../tools/TestTool');
const ReportTool = require('../tools/ReportTool');
const RollbackTool = require('../tools/RollbackTool');

class AutonomousAgent {
  constructor() {
    // Initialize Claude
    this.anthropic = new Anthropic({
      apiKey: process.env.ANTHROPIC_API_KEY
    });
    
    // Initialize core components
    this.stateManager = new StateManager();
    this.decisionEngine = new DecisionEngine(this);
    this.errorHandler = new ErrorHandler(this);
    
    // Initialize tools
    this.tools = {
      search: new SearchTool(this),
      analyze: new AnalyzeTool(this),
      organize: new OrganizeTool(this),
      transform: new TransformTool(this),
      translate: new TranslateTool(this),
      locale: new LocaleTool(this),
      setup: new SetupTool(this),
      integrate: new IntegrateTool(this),
      validate: new ValidateTool(this),
      test: new TestTool(this),
      report: new ReportTool(this),
      rollback: new RollbackTool(this)
    };
    
    // Configuration
    this.config = {
      maxIterations: 50,
      timeout: 300000, // 5 minutes
      targetLanguage: process.env.I18N_TARGET_LANGUAGE || 'es'
    };
    
    // No need for targets file - files are stored in memory
  }

  async start() {
    console.log('🤖 Starting Autonomous I18n Agent...');
    console.log('='.repeat(60));
    
    try {
      // Initialize
      await this.initialize();
      
      // Main agent loop
      const success = await this.runMainLoop();
      
      if (success) {
        console.log('\n🎉 Agent completed successfully!');
        return true;
      } else {
        console.log('\n❌ Agent failed to complete the task');
        return false;
      }
      
    } catch (error) {
      console.error('\n💥 Agent crashed:', error.message);
      await this.emergencyRollback();
      return false;
    }
  }

  async initialize() {
    console.log('🔧 Initializing agent...');
    
    // Set backup directory
    this.backupDir = path.join(process.cwd(), 'backups', `i18n-${Date.now()}`);
    
    // Set target language
    this.stateManager.updateState({
      targetLanguage: this.config.targetLanguage,
      phase: 'initializing'
    });
    
    // Create backup
    await this.tools.rollback.createBackup();
    
    // Set initial goal
    this.stateManager.updateState({
      currentGoal: 'Internationalize the React application',
      phase: 'analyzing'
    });
    
    console.log('✅ Initialization complete');
  }

  async runMainLoop() {
    const startTime = Date.now();
    let iterations = 0;
    
    while (!this.stateManager.shouldStop() && iterations < this.config.maxIterations) {
      // Check timeout
      if (Date.now() - startTime > this.config.timeout) {
        console.log('⏰ Agent timeout reached');
        break;
      }
      
      iterations++;
      console.log(`\n🔄 Iteration ${iterations}/${this.config.maxIterations}`);
      
      try {
        // Decide what to do next
        const decision = await this.decisionEngine.decideNextAction();
        
        // Execute the decision
        const result = await this.executeAction(decision);
        
        // Update state based on result
        this.updateStateFromResult(decision.action, result);
        
        // Check if we're done
        if (this.isGoalAchieved()) {
          this.stateManager.setPhase('completed');
          break;
        }
        
      } catch (error) {
        console.error(`❌ Iteration ${iterations} failed:`, error.message);
        
        // Handle the error
        const errorResult = await this.errorHandler.handleError('main_loop', error);
        
        if (errorResult.stop) {
          console.log('🛑 Agent stopping due to critical error');
          break;
        }
        
        if (errorResult.action === 'rollback') {
          await this.tools.rollback.execute();
          break;
        }
      }
    }
    
    return this.stateManager.isGoalAchieved();
  }

  async executeAction(decision) {
    const { action, reasoning, confidence } = decision;
    
    console.log(`🎯 Executing: ${action} (confidence: ${confidence})`);
    console.log(`💭 Reasoning: ${reasoning}`);
    
    try {
      switch (action) {
        case 'search':
          return await this.executeSearch();
        
        case 'analyze':
          return await this.executeAnalyze();
        
        case 'organize':
          return await this.executeOrganize();
        
        case 'transform':
          return await this.executeTransform();
        
        case 'translate':
          return await this.executeTranslate();
        
        case 'locale':
          return await this.executeLocale();
        
        case 'setup':
          return await this.executeSetup();
        
        case 'integrate':
          return await this.executeIntegrate();
        
        case 'validate':
          return await this.executeValidate();
        
        case 'test':
          return await this.executeTest();
        
        case 'report':
          return await this.executeReport();
        
        case 'retry':
          return await this.executeRetry();
        
        case 'rollback':
          return await this.executeRollback();
        
        case 'complete':
          return await this.executeComplete();
        
        default:
          throw new Error(`Unknown action: ${action}`);
      }
    } catch (error) {
      // Handle tool-specific errors
      return await this.errorHandler.handleError(action, error, { decision });
    }
  }

  async executeSearch() {
    this.stateManager.setPhase('searching');
    const result = await this.tools.search.execute();
    this.stateManager.updateState({ searchResults: result });
    this.stateManager.addCompletedTask('search');
    return result;
  }

  async executeAnalyze() {
    this.stateManager.setPhase('analyzing');
    const result = await this.tools.analyze.execute();
    this.stateManager.updateState({ analysisResults: result });
    this.stateManager.addCompletedTask('analyze');
    return result;
  }

  async executeOrganize() {
    this.stateManager.setPhase('organizing');
    const result = await this.tools.organize.execute();
    this.stateManager.addCompletedTask('organize');
    return result;
  }

  async executeTransform() {
    this.stateManager.setPhase('transforming');
    const result = await this.tools.transform.execute();
    this.stateManager.updateState({ transformResults: result });
    this.stateManager.addCompletedTask('transform');
    return result;
  }

  async executeTranslate() {
    this.stateManager.setPhase('translating');
    const result = await this.tools.translate.execute();
    this.stateManager.updateState({ translateResults: result });
    this.stateManager.addCompletedTask('translate');
    return result;
  }

  async executeLocale() {
    this.stateManager.setPhase('locale');
    const result = await this.tools.locale.execute();
    this.stateManager.updateState({ localeResults: result });
    this.stateManager.addCompletedTask('locale');
    return result;
  }

  async executeSetup() {
    this.stateManager.setPhase('setup');
    const result = await this.tools.setup.execute();
    this.stateManager.updateState({ setupResults: result });
    this.stateManager.addCompletedTask('setup');
    return result;
  }

  async executeIntegrate() {
    this.stateManager.setPhase('integration');
    const result = await this.tools.integrate.execute();
    this.stateManager.updateState({ integrateResults: result });
    this.stateManager.addCompletedTask('integrate');
    return result;
  }

  async executeValidate() {
    this.stateManager.setPhase('validating');
    const result = await this.tools.validate.execute();
    this.stateManager.updateState({ validateResults: result });
    this.stateManager.addCompletedTask('validate');
    return result;
  }

  async executeTest() {
    this.stateManager.setPhase('testing');
    const result = await this.tools.test.execute();
    this.stateManager.addCompletedTask('test');
    return result;
  }

  async executeReport() {
    this.stateManager.setPhase('reporting');
    const result = await this.tools.report.execute();
    this.stateManager.addCompletedTask('report');
    return result;
  }

  async executeRetry() {
    console.log('🔄 Retrying previous failed action...');
    // The decision engine will determine what to retry
    return { action: 'retry', success: true };
  }

  async executeRollback() {
    console.log('🔄 Rolling back changes...');
    const result = await this.tools.rollback.execute();
    this.stateManager.setPhase('failed');
    return result;
  }

  async executeComplete() {
    console.log('✅ Marking task as complete...');
    this.stateManager.setPhase('completed');
    return { action: 'complete', success: true };
  }

  updateStateFromResult(action, result) {
    // Update context based on results
    if (result && result.files) {
      // Handle both array and object formats
      const files = Array.isArray(result.files) ? result.files : Object.keys(result.files);
      this.stateManager.updateContext({
        filesToProcess: files
      });
    }
    
    if (result && result.stringsFound) {
      this.stateManager.updateContext({
        stringsFound: result.stringsFound
      });
    }
    
    if (result && result.issues) {
      this.stateManager.updateContext({
        issues: result.issues
      });
    }
  }

  isGoalAchieved() {
    const state = this.stateManager.getState();
    
    // Check if we have completed the main workflow
    const hasTransformed = state.transformResults && state.transformResults.filesTransformed > 0;
    const hasTranslated = state.translateResults && state.translateResults.stringsTranslated > 0;
    const hasValidated = state.validateResults && state.validateResults.validFiles > 0;
    
    return hasTransformed && hasTranslated && hasValidated;
  }

  async emergencyRollback() {
    console.log('🚨 Emergency rollback initiated...');
    try {
      await this.tools.rollback.execute();
      console.log('✅ Emergency rollback completed');
    } catch (error) {
      console.error('❌ Emergency rollback failed:', error.message);
    }
  }

  // Get agent status for debugging
  getStatus() {
    const state = this.stateManager.getState();
    const errorSummary = this.errorHandler.getErrorSummary();
    
    return {
      phase: state.phase,
      completedTasks: state.completedTasks.length,
      context: state.context,
      errors: errorSummary,
      memory: this.stateManager.getMemoryInsights()
    };
  }
}

module.exports = AutonomousAgent;

File: pythonAgent/core/autonomous_agent.py
"""
Autonomous I18n Agent - Main orchestrator for internationalization workflow.
"""

import asyncio
import os
import logging
from datetime import datetime
from pathlib import Path
from typing import Dict, Any, Optional

from anthropic import Anthropic

from state.state_manager import StateManager
from decision.decision_engine import DecisionEngine
from error.error_handler import ErrorHandler


class AutonomousAgent:
    """Main autonomous agent for internationalizing React applications."""
    
    def __init__(self):
        """Initialize the autonomous agent."""
        # Initialize Claude
        self.anthropic = Anthropic(
            api_key=os.getenv('ANTHROPIC_API_KEY')
        )
        
        # Initialize core components
        self.state_manager = StateManager()
        self.decision_engine = DecisionEngine(self)
        self.error_handler = ErrorHandler(self)
        
        # Initialize tools (will be imported when needed to avoid circular imports)
        self.tools = {}
        
        # Configuration
        self.config = {
            'max_iterations': 50,
            'timeout': 300,  # 5 minutes in seconds
            'target_language': os.getenv('I18N_TARGET_LANGUAGE', 'es')
        }
        
        # Setup logging
        self._setup_logging()
    
    def _setup_logging(self):
        """Setup logging configuration."""
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
            handlers=[
                logging.StreamHandler(),
                logging.FileHandler('i18n_agent.log')
            ]
        )
        # Disable logging for cleaner output
        self.logger = logging.getLogger(__name__)
        self.logger.disabled = True
    
    async def start(self) -> bool:
        """
        Start the autonomous agent.
        
        Returns:
            True if successful, False otherwise
        """
        print('Starting Autonomous I18n Agent...')
        print('=' * 60)
        
        try:
            # Initialize
            await self._initialize()
            
            # Main agent loop
            success = await self._run_main_loop()
            
            if success:
                print('\nSUCCESS: Agent completed successfully!')
                return True
            else:
                print('\nFAILED: Agent failed to complete the task')
                return False
                
        except Exception as error:
            print(f'\nCRASHED: Agent crashed: {error}')
            await self._emergency_rollback()
            return False
    
    async def _initialize(self):
        """Initialize the agent."""
        print('Initializing agent...')
        
        # Set target language
        self.state_manager.update_state({
            'target_language': self.config['target_language'],
            'phase': 'initializing'
        })
        
        # Initialize tools
        await self._initialize_tools()
        
        # Set initial goal
        self.state_manager.update_state({
            'current_goal': 'Internationalize the React application',
            'phase': 'analyzing'
        })
        
        print('SUCCESS: Initialization complete')
    
    async def _initialize_tools(self):
        """Initialize all tools (lazy import to avoid circular dependencies)."""
        # Import tools here to avoid circular imports
        from tools.search_tool import SearchTool
        from tools.analyze_tool import AnalyzeTool
        from tools.transform_tool import TransformTool
        from tools.translate_tool import TranslateTool
        from tools.locale_tool import LocaleTool
        from tools.setup_tool import SetupTool
        from tools.integrate_tool import IntegrateTool
        from tools.validate_tool import ValidateTool
        from tools.rollback_tool import RollbackTool
        from tools.report_tool import ReportTool
        
        # Initialize tools
        self.tools = {
            'search': SearchTool(self),
            'analyze': AnalyzeTool(self),
            'transform': TransformTool(self),
            'translate': TranslateTool(self),
            'locale': LocaleTool(self),
            'setup': SetupTool(self),
            'integrate': IntegrateTool(self),
            'validate': ValidateTool(self),
            'rollback': RollbackTool(self),
            'report': ReportTool(self)
        }
    
    async def _run_main_loop(self) -> bool:
        """Run the main agent loop."""
        start_time = datetime.now()
        iterations = 0
        
        while not self.state_manager.should_stop() and iterations < self.config['max_iterations']:
            # Check timeout
            elapsed = (datetime.now() - start_time).total_seconds()
            if elapsed > self.config['timeout']:
                print('TIMEOUT: Agent timeout reached')
                break
            
            iterations += 1
            print(f'\nIteration {iterations}/{self.config["max_iterations"]}')
            
            try:
                # Decide what to do next
                decision = await self.decision_engine.decide_next_action()
                
                # Execute the decision
                result = await self._execute_action(decision)
                
                # Update state based on result
                self._update_state_from_result(decision.action, result)
                
                # Check if we're done
                if self._is_goal_achieved():
                    self.state_manager.set_phase('completed')
                    break
                
            except KeyboardInterrupt:
                print('\nSTOPPED: Agent stopped by user')
                return False
            except Exception as error:
                print(f'FAILED: Iteration {iterations} failed: {error}')
                
                # Handle the error
                error_result = await self.error_handler.handle_error('main_loop', error)
                
                if error_result.stop:
                    print('🛑 Agent stopping due to critical error')
                    self.logger.error('Agent stopping due to critical error')
                    break
                
                if error_result.action == 'rollback':
                    if 'rollback' in self.tools:
                        await self.tools['rollback'].execute()
                    break
        
        return self.state_manager.is_goal_achieved()
    
    async def _execute_action(self, decision) -> Any:
        """Execute an action based on the decision."""
        action = decision.action
        reasoning = decision.reasoning
        confidence = decision.confidence
        
        print(f'Executing: {action} (confidence: {confidence})')
        print(f'Reasoning: {reasoning}')
        
        try:
            if action == 'search':
                return await self._execute_search()
            elif action == 'analyze':
                return await self._execute_analyze()
            elif action == 'transform':
                return await self._execute_transform()
            elif action == 'translate':
                return await self._execute_translate()
            elif action == 'locale':
                return await self._execute_locale()
            elif action == 'setup':
                return await self._execute_setup()
            elif action == 'integrate':
                return await self._execute_integrate()
            elif action == 'validate':
                return await self._execute_validate()
            elif action == 'retry':
                return await self._execute_retry()
            elif action == 'rollback':
                return await self._execute_rollback()
            elif action == 'complete':
                return await self._execute_complete()
            else:
                raise ValueError(f'Unknown action: {action}')
        except Exception as error:
            # Handle tool-specific errors
            return await self.error_handler.handle_error(action, error, {'decision': decision})
    
    async def _execute_search(self):
        """Execute search action."""
        self.state_manager.set_phase('searching')
        result = await self.tools['search'].execute()
        self.state_manager.update_state({'search_results': result})
        self.state_manager.add_completed_task('search')
        return result
    
    async def _execute_analyze(self):
        """Execute analyze action."""
        self.state_manager.set_phase('analyzing')
        result = await self.tools['analyze'].execute()
        self.state_manager.update_state({'analysis_results': result})
        self.state_manager.add_completed_task('analyze')
        return result
    
    async def _execute_transform(self):
        """Execute transform action."""
        self.state_manager.set_phase('transforming')
        result = await self.tools['transform'].execute()
        self.state_manager.update_state({'transform_results': result})
        self.state_manager.add_completed_task('transform')
        return result
    
    async def _execute_translate(self):
        """Execute translate action."""
        self.state_manager.set_phase('translating')
        result = await self.tools['translate'].execute()
        self.state_manager.update_state({'translate_results': result})
        self.state_manager.add_completed_task('translate')
        return result
    
    async def _execute_locale(self):
        """Execute locale action."""
        self.state_manager.set_phase('locale')
        result = await self.tools['locale'].execute()
        self.state_manager.update_state({'locale_results': result})
        self.state_manager.add_completed_task('locale')
        return result
    
    async def _execute_setup(self):
        """Execute setup action."""
        self.state_manager.set_phase('setup')
        result = await self.tools['setup'].execute()
        self.state_manager.update_state({'setup_results': result})
        self.state_manager.add_completed_task('setup')
        return result
    
    async def _execute_integrate(self):
        """Execute integrate action."""
        self.state_manager.set_phase('integration')
        result = await self.tools['integrate'].execute()
        self.state_manager.update_state({'integrate_results': result})
        self.state_manager.add_completed_task('integrate')
        return result
    
    async def _execute_validate(self):
        """Execute validate action."""
        self.state_manager.set_phase('validating')
        result = await self.tools['validate'].execute()
        self.state_manager.update_state({'validate_results': result})
        self.state_manager.add_completed_task('validate')
        return result
    
    async def _execute_retry(self):
        """Execute retry action."""
        print('Retrying previous action...')
        self.logger.info('Retrying previous action')
        # Simple retry - just continue with the loop
        return {'retry': True}
    
    async def _execute_rollback(self):
        """Execute rollback action."""
        if 'rollback' in self.tools:
            return await self.tools['rollback'].execute()
        return {'rollback': True}
    
    async def _execute_complete(self):
        """Execute complete action."""
        print('SUCCESS: Marking task as complete...')
        self.logger.info('Marking task as complete')
        self.state_manager.set_phase('completed')
        return {'complete': True}
    
    def _update_state_from_result(self, action: str, result: Any):
        """Update state based on action result."""
        if result and isinstance(result, dict):
            if 'files_found' in result:
                self.state_manager.update_context({'files_to_process': result['files_found']})
            if 'strings_found' in result:
                self.state_manager.update_context({'strings_found': result['strings_found']})
            if 'issues' in result:
                self.state_manager.update_context({'issues': result['issues']})
    
    def _is_goal_achieved(self) -> bool:
        """Check if the goal has been achieved."""
        return self.state_manager.is_goal_achieved()
    
    async def _emergency_rollback(self):
        """Perform emergency rollback."""
        print('EMERGENCY: Performing emergency rollback...')
        self.logger.error('Performing emergency rollback')
        
        if 'rollback' in self.tools:
            try:
                await self.tools['rollback'].execute()
            except Exception as error:
                print(f'FAILED: Emergency rollback failed: {error}')
                self.logger.error(f'Emergency rollback failed: {error}')
    
    def get_status(self) -> Dict[str, Any]:
        """Get current agent status."""
        state = self.state_manager.get_state()
        return {
            'phase': state.phase,
            'target_language': state.target_language,
            'completed_tasks': len(state.completed_tasks),
            'files_processed': len(state.context.files_to_process),
            'strings_found': state.context.strings_found if state.context.strings_found else 0,
            'issues': len(state.context.issues),
            'is_goal_achieved': self._is_goal_achieved()
        }
